File: src/importers/column_detector.py
"""Locates the Tally TB header block by searching for known marker text rather than
assuming a fixed row/column position (source plan 6.1 requirement: flexible column
detection since Tally exports may shift slightly)."""

from dataclasses import dataclass
from openpyxl.worksheet.worksheet import Worksheet


@dataclass
class ColumnLayout:
    particulars_col: int
    opening_col: int
    period_debit_col: int
    period_credit_col: int
    closing_col: int
    data_start_row: int


def _norm(value) -> str:
    return str(value).strip().lower() if value is not None else ""


def detect(ws: Worksheet) -> ColumnLayout:
    particulars_row = None
    particulars_col = None
    for row in range(1, min(ws.max_row, 30) + 1):
        for col in range(1, min(ws.max_column, 15) + 1):
            if _norm(ws.cell(row=row, column=col).value) == "particulars":
                particulars_row, particulars_col = row, col
                break
        if particulars_row:
            break
    if particulars_row is None:
        raise ValueError("Could not locate a 'Particulars' header cell in this workbook")

    opening_col = period_start_col = closing_col = None
    marker_row = particulars_row + 1
    for col in range(1, min(ws.max_column, 20) + 1):
        val = _norm(ws.cell(row=marker_row, column=col).value)
        if val == "opening":
            opening_col = col
        elif val == "transactions":
            period_start_col = col
        elif val == "closing":
            closing_col = col
    if not all([opening_col, period_start_col, closing_col]):
        raise ValueError(
            f"Could not locate Opening/Transactions/Closing markers on row {marker_row}"
        )

    debit_col = credit_col = None
    sub_marker_row = particulars_row + 2
    for col in range(period_start_col, closing_col):
        val = _norm(ws.cell(row=sub_marker_row, column=col).value)
        if val == "debit":
            debit_col = col
        elif val == "credit":
            credit_col = col
    if not all([debit_col, credit_col]):
        raise ValueError(
            f"Could not locate Debit/Credit markers on row {sub_marker_row}"
        )

    return ColumnLayout(
        particulars_col=particulars_col,
        opening_col=opening_col,
        period_debit_col=debit_col,
        period_credit_col=credit_col,
        closing_col=closing_col,
        data_start_row=particulars_row + 3,
    )

File: docs/src/excel/column-detector.js
// Port of src/importers/column_detector.py — locates the Tally TB header block by
// searching for known marker text rather than assuming a fixed row/column position.

function norm(value) {
  if (value === null || value === undefined) return "";
  return String(value).trim().toLowerCase();
}

function cellAt(ws, row, col) {
  return ws.getRow(row).getCell(col);
}

export function detect(ws) {
  const maxRow = ws.rowCount;
  const maxCol = ws.columnCount;

  let particularsRow = null;
  let particularsCol = null;
  outer: for (let row = 1; row <= Math.min(maxRow, 30); row++) {
    for (let col = 1; col <= Math.min(maxCol, 15); col++) {
      if (norm(cellAt(ws, row, col).value) === "particulars") {
        particularsRow = row;
        particularsCol = col;
        break outer;
      }
    }
  }
  if (particularsRow === null) {
    throw new Error("Could not locate a 'Particulars' header cell in this workbook");
  }

  // ExcelJS mirrors a merged cell's value across every cell in the merge range (openpyxl
  // only populates the anchor cell, leaving the rest None) — so a match must only be taken
  // the first time, otherwise a two-column "Transactions" merge overwrites periodStartCol
  // with its second half instead of its (correct) anchor/leftmost column.
  let openingCol = null;
  let periodStartCol = null;
  let closingCol = null;
  const markerRow = particularsRow + 1;
  for (let col = 1; col <= Math.min(maxCol, 20); col++) {
    const val = norm(cellAt(ws, markerRow, col).value);
    if (val === "opening" && openingCol === null) openingCol = col;
    else if (val === "transactions" && periodStartCol === null) periodStartCol = col;
    else if (val === "closing" && closingCol === null) closingCol = col;
  }
  if (!openingCol || !periodStartCol || !closingCol) {
    throw new Error(`Could not locate Opening/Transactions/Closing markers on row ${markerRow}`);
  }

  let debitCol = null;
  let creditCol = null;
  const subMarkerRow = particularsRow + 2;
  for (let col = periodStartCol; col < closingCol; col++) {
    const val = norm(cellAt(ws, subMarkerRow, col).value);
    if (val === "debit" && debitCol === null) debitCol = col;
    else if (val === "credit" && creditCol === null) creditCol = col;
  }
  if (!debitCol || !creditCol) {
    throw new Error(`Could not locate Debit/Credit markers on row ${subMarkerRow}`);
  }

  return {
    particularsCol,
    openingCol,
    periodDebitCol: debitCol,
    periodCreditCol: creditCol,
    closingCol,
    dataStartRow: particularsRow + 3,
  };
}
